### ISSUE TITLE
ArrayAccess expressions on primitive array types results in NullReferenceException

### ISSUE THREAD

**Issue by hannasm** [EamonNerbonne/ExpressionToCode#50]
Demonstrated by the following unit test

```csharp
        [Test]
        public void Test002()
        {
            var param = Expression.Parameter(typeof(string).MakeArrayType(), "a");
            var expr = Expression.Lambda(
                Expression.ArrayAccess(param, Expression.Constant(1)),
                param
            );
            Assert.AreEqual(
                "a => a[1]",
                ExpressionToCode.ToCode(expr)
            );
        }
```

Output is:

```
System.NullReferenceException : Object reference not set to an instance of an object.
   at ExpressionToCodeLib.ExpressionToCodeImpl.DispatchIndex(Expression e) in ExpressionToCodeImpl.cs: line 349
   at ExpressionToCodeLib.ExpressionTypeDispatcher.ExpressionDispatch(IExpressionTypeDispatch dispatcher, Expression e) in ExpressionTypeDispatch.Generated.cs: line 152
   at ExpressionToCodeLib.ExpressionToCodeImpl.RawChildDispatch(Argument child) in ExpressionToCodeImpl.cs: line 55
   at ExpressionToCodeLib.ExpressionToCodeImpl.RawChildDispatch(Expression child) in ExpressionToCodeImpl.cs: line 47
   at ExpressionToCodeLib.ExpressionToCodeImpl.NestExpression(Nullable`1 parentType, Expression child, Boolean parensIfEqualRank) in ExpressionToCodeImpl.cs: line 41
   at ExpressionToCodeLib.ExpressionToCodeImpl.DispatchLambda(Expression e) in ExpressionToCodeImpl.cs: line 211
   at ExpressionToCodeLib.ExpressionTypeDispatcher.ExpressionDispatch(IExpressionTypeDispatch dispatcher, Expression e) in ExpressionTypeDispatch.Generated.cs: line 115
   at ExpressionToCodeLib.Unstable_v2_Api.ExpressionStringify.ExpressionToCodeLib.Unstable_v2_Api.IExpressionToCode.ToCode(Expression e) in ExpressionStringify.cs: line 23
   at ExpressionToCodeLib.ExpressionToCode.ToCode(Expression e) in ExpressionToCode.cs: line 25
   at ExpressionToCodeTest.ArrayAccessTests.Test002() in ArrayAccessTests.cs: line 32
````

**Comment by EamonNerbonne** [comment]
Nice find!

I commented on a few minor coding style issues in the PR.

**Comment by hannasm** [comment]
Hi, you might be interested to hear about my use case for expression to code as well. 

I have been generating expression trees and rendering them to c# for the purposes of code generation. The expression tree API provides a robust way to define C# code  on its own, including runtime  type checking, and then Expression<Func<T,...>> takes this even further with compile time type checking and code generation in the same language that you are generating (I use son fancy template functions to deal with substitutions). 

So expression  to code is giving me a way to generate c# code that I can thrn include in my projects, compile and execute.

There are still some gaps here, expression trees do not allow declarations for methods fields classes namespaces using directives, etc... 

I'd like to provide some definitions for these language constructs and leverage them through something like expression to code.

I'm not sure when and where the time for this will be found but if you have some thoughts I'm interested.

-----Original Message-----

**Comment by EamonNerbonne** [comment]
Yeah, that sounds incredibly practical!  Particularly the ability to do substitutions is something I've specifically noticed I'm missing e.g. when implementing [ValueUtils](http://eamonnerbonne.github.io/ValueUtils/).

**Comment by EamonNerbonne** [comment]
I released your changes as version [1.5.3](https://www.nuget.org/packages/ExpressionToCodeLib/1.5.3/).  

Unfortunately, I forgot to explicitly ask: do you license all contributions to this repository under the Apache 2.0 license?

**Comment by hannasm** [comment]
Yes, that is fine. I dont really care to hold any license to that code. I will happily license it however is best for the project.

-----Original Message-----
